File: packages/sdk-python/invoice_ai/money.py
"""
Money helpers. Every amount on the wire is an integer in the currency's
smallest unit (Stripe-style): 2500 is $25.00, 2500 is ¥2,500, 2500 is
KWD 2.500. These helpers convert with each currency's ISO 4217 exponent — the
same table the server uses (lib/currency.ts) — using ``Decimal``, never float
multiplication.
"""

from __future__ import annotations

import re
from decimal import Decimal, InvalidOperation
from typing import Mapping, Union

from typing_extensions import Literal

__all__ = ["currency_decimals", "format_money", "from_minor", "to_minor"]

#: Currencies whose minor unit isn't 1/100. Everything else has 2 decimals.
_DECIMALS: Mapping[str, int] = {
    # Zero-decimal
    **dict.fromkeys(
        [
            "BIF", "CLP", "DJF", "GNF", "ISK", "JPY", "KMF", "KRW", "PYG",
            "RWF", "UGX", "UYI", "VND", "VUV", "XAF", "XOF", "XPF",
        ],
        0,
    ),
    # Three-decimal
    **dict.fromkeys(["BHD", "IQD", "JOD", "KWD", "LYD", "OMR", "TND"], 3),
    # Four-decimal
    "CLF": 4,
    "UYW": 4,
}

_NBSP = chr(0xA0)  # Intl puts a no-break space between a code and the number.
_CODE_RE = re.compile(r"^[A-Z]{3}$")
_AMOUNT_RE = re.compile(r"^([+-]?)(\d*)(?:\.(\d*))?$")

Amount = Union[str, int, float, Decimal]


def currency_decimals(currency: str) -> int:
    """How many decimals a currency uses: USD 2, JPY 0, KWD 3."""
    code = currency.upper()
    if not _CODE_RE.match(code):
        raise ValueError(f"Not an ISO 4217 currency code: {currency}")
    return _DECIMALS.get(code, 2)


def to_minor(amount: Amount, currency: str) -> int:
    """
    A major-unit amount to integer minor units.

        to_minor("25.00", "USD") → 2500     to_minor(5000, "JPY") → 5000
        to_minor("1.234", "KWD") → 1234     to_minor("1.234", "USD") → ValueError

    Raises instead of rounding when the amount has more decimals than the currency.
    """
    decimals = currency_decimals(currency)
    if isinstance(amount, bool):
        raise ValueError(f"Not a decimal amount: {amount!r}")
    if isinstance(amount, Decimal):
        if not amount.is_finite():
            raise ValueError(f"Amount must be finite, got {amount}")
        text = format(amount, "f")
    elif isinstance(amount, int):
        text = str(amount)
    elif isinstance(amount, float):
        if amount != amount or amount in (float("inf"), float("-inf")):
            raise ValueError(f"Amount must be finite, got {amount}")
        # repr() is the shortest round-tripping form (19.99, not 19.989999…).
        text = format(Decimal(repr(amount)), "f")
    elif isinstance(amount, str):
        text = re.sub(r"[,_\s]", "", amount.strip())
    else:
        raise TypeError(f"Amount must be str, int, float or Decimal, got {type(amount).__name__}")

    m = _AMOUNT_RE.match(text)
    if not m or (m.group(2) == "" and (m.group(3) or "") == ""):
        raise ValueError(f"Not a decimal amount: {amount!r}")
    sign, whole, frac_raw = m.group(1), m.group(2), m.group(3) or ""
    frac = frac_raw.rstrip("0")
    if len(frac) > decimals:
        raise ValueError(f"{currency.upper()} allows {decimals} decimal place(s); got {amount}")
    value = int((whole or "0") + frac.ljust(decimals, "0"))
    return -value if sign == "-" and value != 0 else value


def from_minor(minor: int, currency: str) -> Decimal:
    """
    Integer minor units to an exact major-unit ``Decimal``.

        from_minor(2500, "USD") → Decimal("25.00")   from_minor(5000, "JPY") → Decimal("5000")
        from_minor(1234, "KWD") → Decimal("1.234")

    ``str()`` of the result is the exact decimal string (``"25.00"``).
    """
    decimals = currency_decimals(currency)
    if isinstance(minor, bool) or not isinstance(minor, int):
        raise ValueError(f"Minor units must be an integer, got {minor!r}")
    return Decimal(minor).scaleb(-decimals)


#: Symbols as ``Intl.NumberFormat('en-US', {currencyDisplay: 'symbol'})`` prints them.
_SYMBOLS: Mapping[str, str] = {
    "USD": "$",
    "EUR": "€",
    "GBP": "£",
    "JPY": "¥",
    "INR": "₹",
    "CNY": "CN¥",
    "KRW": "₩",
    "CAD": "CA$",
    "AUD": "A$",
    "NZD": "NZ$",
    "HKD": "HK$",
    "MXN": "MX$",
    "BRL": "R$",
    "TWD": "NT$",
    "ILS": "₪",
    "VND": "₫",
    "PHP": "₱",
    "XAF": "FCFA",
    "XOF": "F\u202fCFA",
}


def format_money(
    minor: int,
    currency: str,
    *,
    currency_display: Literal["symbol", "code"] = "symbol",
) -> str:
    """
    Minor units formatted for people (en-US style grouping).

        format_money(2500, "USD") → "$25.00"   format_money(5000, "JPY") → "¥5,000"
        format_money(1234, "KWD") → "KWD 1.234"
    """
    decimals = currency_decimals(currency)
    code = currency.upper()
    major = from_minor(minor, currency)
    try:
        number = f"{abs(major):,.{decimals}f}"
    except (InvalidOperation, ValueError):  # pragma: no cover - Decimal always formats
        number = str(abs(major))
    sign = "-" if major < 0 else ""
    symbol = _SYMBOLS.get(code) if currency_display == "symbol" else None
    if symbol is None:
        return f"{sign}{code}{_NBSP}{number}"
    return f"{sign}{symbol}{number}"

File: packages/sdk-ts/src/core/money.ts
/**
 * Money helpers. Every amount on the wire is an integer in the currency's
 * smallest unit (Stripe-style): 2500 is $25.00, 2500 is ¥2,500, 2500 is
 * KWD 2.500. These helpers convert with each currency's ISO 4217 exponent —
 * the same numbers `Intl` (and therefore the server) uses — without ever
 * going through floating-point multiplication.
 */

/** Currencies whose minor unit isn't 1/100. Everything else has 2 decimals. */
const DECIMALS: Readonly<Record<string, number>> = {
  // Zero-decimal
  BIF: 0, CLP: 0, DJF: 0, GNF: 0, ISK: 0, JPY: 0, KMF: 0, KRW: 0, PYG: 0,
  RWF: 0, UGX: 0, UYI: 0, VND: 0, VUV: 0, XAF: 0, XOF: 0, XPF: 0,
  // Three-decimal
  BHD: 3, IQD: 3, JOD: 3, KWD: 3, LYD: 3, OMR: 3, TND: 3,
  // Four-decimal
  CLF: 4, UYW: 4,
}

/** How many decimals a currency uses: USD 2, JPY 0, KWD 3. */
export function currencyDecimals(currency: string): number {
  const code = currency.toUpperCase()
  if (!/^[A-Z]{3}$/.test(code)) throw new RangeError(`Not an ISO 4217 currency code: ${currency}`)
  return DECIMALS[code] ?? 2
}

/**
 * A major-unit amount to integer minor units.
 *   toMinor('25.00', 'USD') → 2500     toMinor(5000, 'JPY') → 5000
 *   toMinor('1.234', 'KWD') → 1234     toMinor('1.234', 'USD') → RangeError
 * Throws instead of rounding when the amount has more decimals than the currency.
 */
export function toMinor(amount: string | number, currency: string): number {
  const decimals = currencyDecimals(currency)
  let text: string
  if (typeof amount === 'number') {
    if (!Number.isFinite(amount)) throw new RangeError(`Amount must be finite, got ${amount}`)
    text = String(amount)
    if (/e/i.test(text)) text = amount.toFixed(Math.max(decimals, 0))
  } else {
    text = amount.trim().replace(/[,_\s]/g, '')
  }
  const m = text.match(/^([+-]?)(\d*)(?:\.(\d*))?$/)
  if (!m || (m[2] === '' && (m[3] ?? '') === '')) throw new RangeError(`Not a decimal amount: ${String(amount)}`)
  const [, sign, whole = '', fracRaw = ''] = m
  const frac = fracRaw.replace(/0+$/, '')
  if (frac.length > decimals) {
    throw new RangeError(`${currency.toUpperCase()} allows ${decimals} decimal place(s); got ${String(amount)}`)
  }
  const digits = (whole || '0') + frac.padEnd(decimals, '0')
  const value = Number(digits)
  if (!Number.isSafeInteger(value)) throw new RangeError(`Amount is too large: ${String(amount)}`)
  return sign === '-' && value !== 0 ? -value : value
}

/**
 * Integer minor units to an exact major-unit decimal string.
 *   fromMinor(2500, 'USD') → '25.00'   fromMinor(5000, 'JPY') → '5000'   fromMinor(1234, 'KWD') → '1.234'
 */
export function fromMinor(minor: number | bigint, currency: string): string {
  const decimals = currencyDecimals(currency)
  if (typeof minor === 'number' && !Number.isInteger(minor)) {
    throw new RangeError(`Minor units must be an integer, got ${minor}`)
  }
  const n = BigInt(minor)
  const negative = n < BigInt(0)
  const digits = (negative ? -n : n).toString()
  if (decimals === 0) return (negative ? '-' : '') + digits
  const padded = digits.padStart(decimals + 1, '0')
  const whole = padded.slice(0, -decimals)
  const frac = padded.slice(-decimals)
  return `${negative ? '-' : ''}${whole}.${frac}`
}

export interface FormatMoneyOptions {
  /** BCP 47 locale. Default `en-US`. */
  locale?: string
  /** `symbol` ($), `narrowSymbol`, `code` (USD) or `name`. Default `symbol`. */
  currencyDisplay?: 'symbol' | 'narrowSymbol' | 'code' | 'name'
}

/**
 * Minor units formatted for people.
 *   formatMoney(2500, 'USD') → '$25.00'   formatMoney(5000, 'JPY') → '¥5,000'
 */
export function formatMoney(minor: number | bigint, currency: string, options: FormatMoneyOptions = {}): string {
  const decimals = currencyDecimals(currency)
  const major = fromMinor(minor, currency)
  try {
    return new Intl.NumberFormat(options.locale ?? 'en-US', {
      style: 'currency',
      currency: currency.toUpperCase(),
      currencyDisplay: options.currencyDisplay ?? 'symbol',
      minimumFractionDigits: decimals,
      maximumFractionDigits: decimals,
    }).format(Number(major))
  } catch {
    return `${currency.toUpperCase()} ${major}`
  }
}
